src/Microsoft.OData.Client/BulkUpdateGraph.cs

```diff
@@ -0,0 +1,146 @@
+            this.descriptorGraph= new Dictionary<Descriptor, HashSet<Descriptor>>(); 
+            this.topLevelDescriptors= new List<EntityDescriptor>();
+            this.requestInfo = requestInfo;
+        }
+
+        /// <summary>
+        /// Returns a collection of all the root descriptors used in a deep update.
+        /// </summary>
+        public List<EntityDescriptor> TopLevelDescriptors
+        {
+            get
+            {
+                return this.topLevelDescriptors;
+            }
+        }
+
+        /// <summary>
+        /// Returns the EntitySetName.
+        /// </summary>
+        public string EntitySetName
+        {
+            get
+            {
+                return this.entitySetName;
+            }
+        }
+
+        /// <summary>
+        /// Computes the EntyitySetName in cases where the top-level EntityDescriptors
+        /// do not have the EntitySetName set. 
+        /// </summary>
+        public void ComputeEntitySetName()
+        {
+            EntityDescriptor parentDescriptor = this.TopLevelDescriptors[0] as EntityDescriptor;
+
+            if (!string.IsNullOrEmpty(parentDescriptor.EntitySetName))
+            {
+                this.entitySetName = parentDescriptor.EntitySetName;
+            }
+            else
+            {
+                Uri baseUri = new Uri(this.requestInfo.BaseUriResolver.GetBaseUriWithSlash().ToString());
+                Uri resourceUri = parentDescriptor.EditLink;
+                ODataUriParser parser = new ODataUriParser(this.requestInfo.Format.ServiceModel, baseUri, resourceUri);
+                ODataPath path = parser.ParsePath();
+
+                if (path != null)
+                {
+                    EntitySetSegment entitySetSegment = path.ToList().OfType<EntitySetSegment>().LastOrDefault();
+                    this.entitySetName = entitySetSegment.Identifier;
+                }
+                else
+                {
+                    throw Error.InvalidOperation(Strings.DataBinding_Util_UnknownEntitySetName(parentDescriptor.Entity.GetType().FullName));
+                }
+            }
+        }
+
+        /// <summary>
+        /// Links a descriptor with its related descriptors.
+        /// </summary>
+        /// <param name="parent">The parent descriptor.</param>
+        /// <param name="child">The related descriptor to the parent descriptor.</param>
+        public void AddRelatedDescriptor(Descriptor parent, Descriptor child)
+        {
+            HashSet<Descriptor> childrenDescriptors = this.GetRelatedDescriptors(parent);
+            childrenDescriptors.Add(child);
+  
+        }
+
+        /// <summary>
+        /// Adds to the top-level descriptors' list.
+        /// </summary>
+        /// <param name="descriptor">The descriptor to be added to the top-level descriptors' list.</param>
+        public void AddTopLevelDescriptor(EntityDescriptor descriptor)
+        {
+            if (this.topLevelDescriptors == null)

```

Comment:
Since you initialize the topLevelDescriptors in the constructor, line 109 is never working?